presto-main/src/main/java/com/facebook/presto/execution/executor/TaskExecutor.java

```diff
@@ -193,16 +210,225 @@ public TaskExecutor(TaskManagerConfig config, EmbedVersion embedVersion, Multile
                 DEFAULT_INTERRUPT_SPLIT_INTERVAL,
                 embedVersion,
                 splitQueue,
+                gracefulShutdownSplitTracker,
                 Ticker.systemTicker());
     }
 
+    public void gracefulShutdown()
+    {
+        isGracefulShutdownStarted.set(true);
+        long shutdownStartTime = System.nanoTime();
+        //TODO throw error for new task creation instead of looping here.
+        List<TaskHandle> activeTaskSnapshot = getActiveTasks();
+        while (activeTaskSnapshot.size() > 0) {
+            gracefulShutdown(activeTaskSnapshot);
+            activeTaskSnapshot = getActiveTasks();
+        }
```

Comment:
`gracefulShutdown(activeTaskSnapshot);` is async and will return early. 
What happens if `gracefulShutdown(activeTaskSnapshot)` is called multiple times on the same elements. Is it idempotent?